### PR TITLE
Support structural mass matrix in HMC

### LINKED ISSUE #957: Redesign kinetic energy api to allow more flexible structural mass matrix
reference: [BlockMassMatrix](http://docs.pyro.ai/en/stable/mcmc.html#blockmassmatrix) in pyro

cc @martinjankowiak

### LINKED ISSUE #961: Allow users to set inverse mass matrix for initial state
This feature has been requested by several users.

Currently, we can only do this with some understanding of NumPyro codebase. But it would be nice to have an API that is more intuitive for users. The tricky thing is the notion of `inverse_mass_matrix` is a bit ambiguous (w.r.t. the order of latent variables).

### PR DESCRIPTION
**Pull request by fehiepsi**
Fixes #957 and fixes #961.

Previously, I thought it is very complicated to support this feature. Although we already have this in Pyro, the object orient pattern there is quite different from NumPyro. After several tries (attempts to have a general object failed - due to a lot of unmanageable changes in codebase), I come up with the following solution that is pretty clean!! :D

### What this PR does
+ support structural mass matrix with the pattern {tuple-sites: mass-matrix-value}
+ mass matrix values can be diagonal or dense
+ the order of entries in a mass matrix is governed by the order in a tuple of sites
+ with the above, users can specify the initial mass matrix, without ambiguity

### TODO
- [x] implement the core logic
- [x] update Welford covariance to use structural mass
- [x] update warmup adapter to account for structural updates
- [x] fixes the remaining usages of inv mass matrix
- [x] allow users to specify mass matrix in the HMC kernel
- [x] add tests

### CODE REVIEW

File: numpyro/infer/hmc.py [hunk 1]
```diff
@@ -374,12 +382,38 @@ class HMC(MCMCKernel):
     :param float step_size: Determines the size of a single step taken by the
         verlet integrator while computing the trajectory using Hamiltonian
         dynamics. If not specified, it will be set to 1.
+    :param inverse_mass_matrix: Initial value for inverse mass matrix.
+        This may be adapted during warmup if adapt_mass_matrix = True.
+        If no value is specified, then it is initialized to the identity matrix.
+        For a potential_fn with general JAX pytree parameters, the order of entries
+        of the mass matrix is the order of the flatten version of pytree parameters,
```
Comment (martinjankowiak) [inline numpyro/infer/hmc.py:389]: nit, is this better?  

... of the flattened version of pytree parameters obtained with `tree_flatten`,

File: numpyro/infer/hmc.py [hunk 2]
```diff
@@ -374,12 +382,38 @@ class HMC(MCMCKernel):
...
+        which is a bit ambiguous (see more at
+        https://jax.readthedocs.io/en/latest/pytrees.html). If `model` is not None,
+        here we can specify a structured block mass matrix, with keys are tuple of
+        site names and values are the corresponding block of the mass matrix.
+        For more information about structured mass matrix, see `dense_mass` argument.
```
Comment (martinjankowiak) [inline numpyro/infer/hmc.py:394]: nit:

a structured block mass matrix as a dictionary, where keys are tuples of
site names and values are the corresponding block of the mass matrix.
For more information about structured mass matrices, ...

File: numpyro/infer/hmc.py [hunk 3]
```diff
@@ -374,12 +382,38 @@ class HMC(MCMCKernel):
...
+    :type inverse_mass_matrix: numpy.ndarray or dict
     :param bool adapt_step_size: A flag to decide if we want to adapt step_size
         during warm-up phase using Dual Averaging scheme.
     :param bool adapt_mass_matrix: A flag to decide if we want to adapt mass
         matrix during warm-up phase using Welford scheme.
-    :param bool dense_mass:  A flag to decide if mass matrix is dense or
-        diagonal (default when ``dense_mass=False``)
+    :param dense_mass:  A flag to decide if mass matrix is dense or
```
Comment (martinjankowiak) [inline numpyro/infer/hmc.py:400]: nit:

    :param dense_mass:  This flag controls whether mass matrix is dense (i.e. full-rank) or
        diagonal (defaults to ``dense_mass=False``). To specify a structured mass matrix,
        users can provide a list of tuples of site names. Each tuple represents
        a block in the joint mass matrix. For example, assuming that the model
        has latent variables "x", "y", "z" (where each variable can be multi-dimensional),
        possible specifications and corresponding mass matrix structures are as follows:

            + dense_mass=[("x", "y")]: use a dense mass matrix for the joint
              (x, y) and a diagonal mass matrix for z
            + dense_mass=[] (equivalent to dense_mass=False): use a diagonal mass
              matrix for the joint (x, y, z)
            + dense_mass=[("x", "y", "z")] (equivalent to full_mass=True):
              use a dense mass matrix for the joint (x, y, z)
            + dense_mass=[("x",), ("y",), ("z")]: use dense mass matrices for
              each of x, y, and z (i.e. block-diagonal with 3 blocks)

File: numpyro/infer/hmc.py [hunk 4]
```diff
@@ -485,13 +521,31 @@ def init(self, rng_key, num_warmup, init_params=None, model_args=(), model_kwarg
...
+                # this is to be compatible with older numpyro versions
+                # and to match autoguide scale parameter and jax flatten utils
+                dense_mass = [tuple(sorted(z))] if dense_mass else []
+            else:
+                # assert that site_names in dense_mass and user-defined
+                # inverse_mass_matrix matched
+                assert isinstance(dense_mass, list)
+                if isinstance(self._inverse_mass_matrix, dict):
+                    for site_names in dense_mass:
+                        assert site_names in self._inverse_mass_matrix
```
Comment (martinjankowiak) [inline numpyro/infer/hmc.py:539]: + error message?

File: numpyro/infer/hmc.py [hunk 5]
```diff
@@ -556,12 +610,38 @@ class NUTS(HMC):
...
+        here are the meaning of various structures:
+
+            + dense_mass=[("x", "y")]: using dense mass matrix for the joint
+              (x, y), diagonal mass matrix for z
+            + dense_mass=[] (equivalent to dense_mass=False): using diagonal mass
+              matrix for the joint (x, y, z)
+            + dense_mass=[("x", "y", "z")] (equivalent to full_mass=True):
+              using dense mass matrix for the joint (x, y, z)
+            + dense_mass=[("x",), ("y",), ("z")]: using dense mass matrices for
+              each of x, y, z
```
Comment (martinjankowiak) [inline numpyro/infer/hmc.py:642]: as above...

File: numpyro/infer/hmc_util.py [hunk 6]
```diff
@@ -118,14 +118,24 @@ def welford_covariance(diagonal=True):
     """
     def init_fn(size):
         """
-        :param int size: size of each sample.
+        :param int size: size of each sample. For structural mass matrix,
```
Comment (martinjankowiak) [inline numpyro/infer/hmc_util.py:121]: ... For a structured mass matrix,...

File: numpyro/infer/hmc_util.py [hunk 7]
```diff
@@ -350,6 +379,65 @@ def build_adaptation_schedule(num_steps):
...
+        for site_names in dense_mass:
+            inverse_mm = inverse_mass_matrix.get(site_names)
+            z_block = tuple(z[k] for k in site_names)
+            inverse_mm, mm_sqrt, mm_sqrt_inv = _initialize_mass_matrix(z_block, inverse_mm, True)
+            inverse_mass_matrix[site_names] = inverse_mm
+            mass_matrix_sqrt[site_names] = mm_sqrt
+            mass_matrix_sqrt_inv[site_names] = mm_sqrt_inv
+        # NB: this branch only happens when users want to use block diagonal
+        # inverse_mass_matrix, for example, {("a",): jnp.ones(3), ("b",): jnp.ones(3)}.
+        for site_names, inverse_mm in inverse_mass_matrix.items():
```
Comment (martinjankowiak) [inline numpyro/infer/hmc_util.py:396]: hmm this part is kind of tricky you may need to walk me through this...

File: test/infer/test_mcmc.py [hunk 8]
```diff
@@ -739,3 +741,60 @@ def model(data):
...
+    cov[:2, :2] = w_cov
+    cov[2:4, 2:4] = xy_cov
+    cov[4, 4] = z_var
+
+    # smoke tests
+    for dense_mass in [True, False]:
+        kernel = NUTS(model, dense_mass=dense_mass)
+        mcmc = MCMC(kernel, num_warmup=1, num_samples=1)
+        mcmc.run(random.PRNGKey(0), cov)
+        actual_dim = mcmc.last_state.adapt_state.inverse_mass_matrix[("w", "x", "y", "z")].ndim
```
Comment (martinjankowiak) [inline test/infer/test_mcmc.py:768]: hmm why does the dict exist if `dense_mass` is a `bool`...

File: test/infer/test_mcmc.py [hunk 9]
```diff
@@ -739,3 +741,60 @@ def model(data):
...
+
+    w_cov = np.array([[1.5, 0.5], [0.5, 1.5]])
+    xy_cov = np.array([[2., 1.], [1., 3.]])
+    z_var = np.array([2.5])
+    cov = np.zeros((5, 5))
+    cov[:2, :2] = w_cov
+    cov[2:4, 2:4] = xy_cov
+    cov[4, 4] = z_var
+
+    # smoke tests
```
Comment (martinjankowiak) [inline test/infer/test_mcmc.py:763]: it might be useful to have a smoketest where the latent variables have different dimensions? e.g. 0, 1, 2

**Review by martinjankowiak — COMMENTED** [review]
this is great!!

File: numpyro/infer/hmc.py (hunk 4 above)
Comment (fehiepsi) [reply numpyro/infer/hmc.py:539]: We will check for consistency in `_initialize_mass_matrix` method and will raise the error there. So I would like to remove this branch.

File: numpyro/infer/hmc_util.py (hunk 7 above)
Comment (fehiepsi) [reply numpyro/infer/hmc_util.py:396]: I guess this is resolved per our discussion.

File: test/infer/test_mcmc.py (hunk 8 above)
Comment (fehiepsi) [reply test/infer/test_mcmc.py:768]: this is resolved via discussion: this is a breaking change, to make inverse_mass_matrix more interpretable.

File: test/infer/test_mcmc.py (hunk 9 above)
Comment (fehiepsi) [reply test/infer/test_mcmc.py]: agreed! I just added a new test to check for the shape and various structured.

File: numpyro/infer/hmc.py [hunk 10]
```diff
@@ -386,32 +386,33 @@ class HMC(MCMCKernel):
         This may be adapted during warmup if adapt_mass_matrix = True.
         If no value is specified, then it is initialized to the identity matrix.
         For a potential_fn with general JAX pytree parameters, the order of entries
-        of the mass matrix is the order of the flatten version of pytree parameters,
-        which is a bit ambiguous (see more at
+        of the mass matrix is the order of the flatten version of pytree parameters
```
Comment (martinjankowiak) [inline numpyro/infer/hmc.py:389]: nit: flattened version

File: numpyro/infer/hmc.py [hunk 11]
```diff
@@ -614,32 +616,33 @@ class NUTS(HMC):
...
```
Comment (martinjankowiak) [inline numpyro/infer/hmc.py:613]: flattened

File: test/infer/test_mcmc.py [hunk 12]
```diff
@@ -785,16 +777,42 @@ def model(cov):
...
-def test_initial_inverse_mass_matrix():
+@pytest.mark.parametrize("dense_mass, expected_shapes", [
+    (False, {("w", "x", "y", "z"): (16,)}),
+    (True, {("w", "x", "y", "z"): (16, 16)}),
+    ([("y", "w", "z", "x")], {("y", "w", "z", "x"): (16, 16)}),
+    ([("x", "w"), ("y",)], {("x", "w"): (11, 11), ("y",): (4, 4), ("z",): (1,)}),
+    ([("y",)], {("w", "x", "z"): (12,), ("y",): (4, 4)}),
+    ([("z",), ("w",), ("y",)], {("w",): (10, 10), ("x",): (1,), ("y",): (4, 4), ("z",): (1, 1)}),
+])
+def test_structured_mass_smoke(dense_mass, expected_shapes):
```
Comment (martinjankowiak) [inline test/infer/test_mcmc.py:788]: 👍

**Review by martinjankowiak — APPROVED** [review]

**Comment by fehiepsi** [comment]
Thank you for reviewing, @martinjankowiak! :)

**Review by martinjankowiak — APPROVED** [review]
